sky/provision/lambda_cloud/instance.py

```diff
@@ -64,6 +64,13 @@ def _get_ssh_key_name(prefix: str = '') -> str:
     return name
 
 
+def _get_private_ip(cluster_name_on_cloud: str,
+                    instance_info: Dict[str, Any]) -> str:
+    if len(_filter_instances(cluster_name_on_cloud, None)) == 1:
+        return instance_info.get('private_ip', '127.0.0.1')
+    return instance_info['private_ip']
```

Comment:
How about we just use a argument called `single_node` for the function, so we don't have to call the cloud API multiple times for initializing the `cluster_info` for better performance? We can set `single_node=(len(running_instances) > 1)`
```suggestion
def _get_private_ip(instance_info: Dict[str, Any], single_node: bool) -> str:
    private_ip = instance_info.get('private_ip')
    if private_ip is None:
        if single_node:
            # The Lambda cloud API may return an instance info without
            # private IP. It does not align with their docs, but we still
            # allow single-node cluster to proceed with the provisioning, by using
            # 127.0.0.1, as private IP is not critical for single-node case.
            return '127.0.0.1'
        msg = f'Failed to retrieve private IP for instance {instance_info}.'
        logger.error(msg)
        raise RuntimeError(msg)
    return private_ip
```

Reply:
I like it and it’s easier to unit test! I introduced the extra API call because I wasn’t sure about the assumption that all nodes of a cluster would be running at the same time but you know better.